File: app/src/main/java/com/application/newsnow/fragment/SectionNewsFeedFragment.java
package com.application.newsnow.fragment;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.application.newsnow.R;
import com.application.newsnow.adapter.NewsAdapter;
import com.application.newsnow.model.ListNews;
import com.application.newsnow.model.News;
import com.application.newsnow.model.Section;
import com.application.newsnow.retrofit.RetrofitInstance;
import com.application.newsnow.util.OnNewsListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SectionNewsFeedFragment extends Fragment implements OnNewsListener {

    private static final String SECTION_KEY_BUNDLE = "section_key";
    private static final String RETURN = "return";
    private static final String FAIL = "Oops, something wrong!";

    private List<News> news = new ArrayList<>();
    private NewsAdapter adapter;
    private Section section;

    private ProgressBar load;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_section_news_feed, container, false);

        section = (Section) getArguments().getSerializable(SECTION_KEY_BUNDLE);
        load = view.findViewById(R.id.load_news_section);

        setToolbar(view);

        initRecyclerView(view);

        generateCall();

        return view;
    }

    private void generateCall() {
        Call<ListNews> call = RetrofitInstance.getInstance()
                .getApi()
                .getAllNews(section.getSection());
        call.enqueue(new Callback<ListNews>() {
            @Override
            public void onResponse(@NonNull Call<ListNews> call, @NonNull Response<ListNews> response) {
                if (response.isSuccessful()) {
                    load.setVisibility(View.GONE);

                    news = response.body().getResults();

                    adapter.addPosters(news);
                }
            }

            @Override
            public void onFailure(@NonNull Call<ListNews> call, @NonNull Throwable t) {
                Toast.makeText(getActivity().getApplicationContext(), FAIL, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public void onNewsClick(News poster) {
        Fragment fragment = NewsDetailFragment.getInstance(poster);

        getActivity().getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.news_fragment_container, fragment, RETURN)
                .addToBackStack(RETURN)
                .commit();
    }

    public static SectionNewsFeedFragment getInstance(Section section) {
        SectionNewsFeedFragment fragment = new SectionNewsFeedFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(SECTION_KEY_BUNDLE, section);
        fragment.setArguments(bundle);
        return fragment;
    }

    private void setToolbar(View view) {
        Toolbar toolbar = view.findViewById(R.id.toolbar_news_section_feed);
        toolbar.inflateMenu(R.menu.menu_top_news);
        toolbar.setTitle(section.getSection().toUpperCase(Locale.ROOT));
        toolbar.setNavigationIcon(R.drawable.ic_baseline_arrow_back_ios_new_24);
        toolbar.setNavigationOnClickListener(viewClick -> getActivity().onBackPressed());
    }

    private void initRecyclerView(View view) {
        RecyclerView recyclerView = view.findViewById(R.id.list_news_section);
        recyclerView.setHasFixedSize(true);

        adapter = new NewsAdapter(this);
        RecyclerView.LayoutManager manager = new LinearLayoutManager(view.getContext());

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(manager);
    }
}

File: app/src/main/java/com/application/newsnow/fragment/SectionNewsFeedFragment.kt
package com.application.newsnow.fragment

import android.content.Context
import android.os.Bundle
import android.view.*
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.application.app.App
import com.application.newsnow.adapter.SearchAdapter
import com.application.newsnow.databinding.FragmentSectionNewsFeedBinding
import com.application.newsnow.fragment.base.BaseFragment
import com.application.newsnow.model.NewsView
import com.application.newsnow.model.Section
import com.application.newsnow.util.OnNewsListener
import com.application.newsnow.viewmodel.SectionNewsViewModel
import com.application.newsnow.viewmodelfactory.NewsViewModelFactory
import javax.inject.Inject

class SectionNewsFeedFragment : BaseFragment<SectionNewsViewModel>(), OnNewsListener {

    companion object {
        private const val SECTION_KEY_BUNDLE = "section_key"

        fun getInstance(section: Section): SectionNewsFeedFragment {
            val fragment = SectionNewsFeedFragment()
            val bundle = Bundle()
            bundle.putSerializable(SECTION_KEY_BUNDLE, section)
            fragment.arguments = bundle
            return fragment
        }
    }

    @Inject
    lateinit var viewModelFactory: NewsViewModelFactory

    override val viewModel by lazy(LazyThreadSafetyMode.NONE) {
        ViewModelProvider(this, viewModelFactory)[SectionNewsViewModel::class.java]
    }

    private val searchAdapter by lazy { SearchAdapter(this) }

    private val section by lazy(LazyThreadSafetyMode.NONE) {
        arguments?.getSerializable(SECTION_KEY_BUNDLE) as Section
    }

    private lateinit var binding: FragmentSectionNewsFeedBinding

    override fun onAttach(context: Context) {
        (activity?.applicationContext as App).component.inject(this)
        super.onAttach(context)
        viewModel.fetchNewsByCategory(section)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentSectionNewsFeedBinding.inflate(inflater, container, false)

        val section = arguments?.getSerializable(SECTION_KEY_BUNDLE) as Section

        setToolbar(section)
        initRecyclerView()
        setSearch()
        fetchNewsByCategory()

        return binding.root
    }

    private fun setToolbar(section: Section) {
        with(binding.toolbarNewsSectionFeed) {
            title = section.section.replaceFirstChar { it.uppercase() }
            setNavigationOnClickListener { requireActivity().onBackPressed() }
        }
    }

    private fun initRecyclerView() {
        with(binding.listNewsSection) {
            setHasFixedSize(true)
            val manager = LinearLayoutManager(context)
            layoutManager = manager
            adapter = searchAdapter
        }
    }

    private fun setSearch() {
        with(binding.searchNews) {
            setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String?): Boolean {
                    return false
                }

                override fun onQueryTextChange(newText: String?): Boolean {
                    searchAdapter.filter.filter(newText)
                    return true
                }
            })
        }
    }

    private fun fetchNewsByCategory() {
        viewModel.liveDataNews.observe(requireActivity()) { response ->
            binding.loadNewsSection.visibility = View.GONE
            searchAdapter.addSearchedNews(response.results)
        }

        viewModel.liveDataException.observe(requireActivity()) { error ->
            binding.loadNewsSection.visibility = View.GONE
            Toast.makeText(context, error, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onNewsClick(news: NewsView) {
        viewModel.navigateToDetails(newsView = news)
    }

}